### DISCUSSION TITLE
RPCs / server$

### DISCUSSION THREAD

**Discussion by Blechlawine** [nuxt/nuxt#20366] [Ideas]
# Background
[Solid Start](https://github.com/solidjs/solid-start) introduced the `server$` function to create callable functions that get compiled to API calls. [ModernJS](https://github.com/web-infra-dev/modern.js) also has a similar feature, allowing the functions exported by API endpoints to be called on the client, which then also get compiled to API calls. There is also [Tanstack/Bling](https://github.com/TanStack/bling) which can add the `server$` function (and more) to any Vite-based project (I couldn't get this working in Nuxt though and they only have examples for Astro with React or Solid).

# My Feature Idea
In Nuxt, API handlers already export a function, but they still have to be called with either `$fetch` or `useFetch`, it would be great if we could directly call the exported functions from the frontend, just like a normal function call, with parameters for the input and correct types all around. Either that, or implement something along the lines of Solid Start or Bling, a separate builder function to create callable endpoints anywhere in the code (even inside components). Although having both might be the best option.

# Why?
This would help tremendously with type-safety across API calls, which currently rely on code-generation for response types. This way we could also get input type-safety and response types **without** code-generation.

**Comment by MattStopa** [comment]
This is the way of the future. I hope this gets some traction.

**Comment by 3xau1o** [comment]
https://github.com/guushamann/Todo-tRPC-Vue3

**Comment by gsxdsm** [comment]
https://github.com/gsxdsm/nuxt-rpc

**Comment by TaQuanMinhLong** [comment]
Checkout usage of orpc with nuxt here https://orpc.unnoq.com/docs/adapters/nuxt
